pkg/terminal/command.go

```diff
@@ -563,35 +570,77 @@ func (c *Commands) goroutine(t *Term, ctx callContext, argstr string) error {
 		if err != nil {
 			return err
 		}
-
+		c.Frame = 0
 		fmt.Printf("Switched from %d to %d (thread %d)\n", selectedGID(oldState), gid, newState.CurrentThread.ID)
 		return nil
 	}
 
 	var err error
-	ctx.Prefix = scopePrefix
 	ctx.Scope.GoroutineID, err = strconv.Atoi(args[0])
 	if err != nil {
 		return err
 	}
 	return c.CallWithContext(args[1], t, ctx)
 }
 
-func (c *Commands) frame(t *Term, ctx callContext, args string) error {
-	v := strings.SplitN(args, " ", 2)
-
-	switch len(v) {
-	case 0, 1:
+func (c *Commands) frame(t *Term, ctx callContext, argstr string) error {
+	if len(argstr) == 0 {
 		return errors.New("not enough arguments")
 	}
+	args := strings.SplitN(argstr, " ", 2)
+	frame, err := strconv.Atoi(args[0])
+	if err != nil {
+		return err
+	}
 
-	var err error
-	ctx.Prefix = scopePrefix
-	ctx.Scope.Frame, err = strconv.Atoi(v[0])
+	if len(args) == 1 {
+		return c.setCurrentFrame(t, ctx, frame)
+	}
+	ctx.Scope.Frame = frame
+	return c.CallWithContext(args[1], t, ctx)
+}
+
+func (c *Commands) setCurrentFrame(t *Term, ctx callContext, frame int) error {
+	if frame < 0 {
+		return fmt.Errorf("Invalid frame %d", frame)
+	}
+	stack, err := t.client.Stacktrace(ctx.Scope.GoroutineID, frame, nil)
 	if err != nil {
 		return err
 	}
-	return c.CallWithContext(v[1], t, ctx)
+	if frame >= len(stack) {
+		return fmt.Errorf("Invalid frame %d", frame)
+	}
+	c.Frame = frame
+	state, err := t.client.GetState()
+	if err != nil {
+		return err
+	}
+	printcontext(t, state)
+	printfile(t, stack[frame].File, stack[frame].Line, true)
+	return nil
+}
+
+func (c *Commands) up(t *Term, ctx callContext, args string) error {
```

Comment:
This and `down` should be implemented by calling to the same function and passing a different "direction" argument.